pennylane/tape/functions/unified.py

```diff
@@ -0,0 +1,210 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Wrappers for common functions that manipulate or create
+NumPy, TensorFlow, and Torch data structures.
+"""
+import abc
+
+
+class UnifiedTensor(abc.ABC):
+    """Creates a unified tensor wrapper, that represents interface array/tensor datatypes
+    using a common API.
+
+    While this is an abstract base class, this class may be 'instantiated' directly;
+    by overloading ``__new__``, the tensor argument is inspected, and the correct subclass
+    is returned:
+
+    >>> x = tf.Variable([0.4, 0.1, 0.5])
+    >>> y = UnifiedTensor(x)
+    >>> type(y)
+    <class 'pennylane.tape.functions.tf.TensorFlowTensor'>

```

Comment:
That is really nice, this is the idea from eagerpy, right?

Reply:
Yep, inspired by their UI! Not sure if this is the same as their implementation though.